Refactor future-post validation process for scalability

ethitter opened the pull request:
Previously, all future posts where requested at once, which didn't always scale. Also, since `post_date_gmt` was requested, there were no indices available. By switching to `post_date` and using Core's `get_gmt_from_date()`, the same data is provided and an index is leveraged.

Fixes #3.

ethitter committed:
Refactor future-post validation to process in smaller batches and use an index WP provides

Previously, all future posts where requested at once, which didn't always scale. Also, since `post_date_gmt` was requested, there were no indicies available. By switching to `post_date` and using Core's `get_gmt_from_date()`, the same data is provided and an index is leveraged.

Fixes #3.

simonwheatley commented:
The core of the loop seems to be:

``` php
                // grab the scheduled job for this post
                $timestamp = wp_next_scheduled( 'publish_future_post', array( (int) $r->ID ) );
                if ( (int) $timestamp !== $gmt_time ) {
                    wp_clear_scheduled_hook( 'publish_future_post', array( (int) $r->ID ) );
                }
                wp_schedule_single_event( $gmt_time, 'publish_future_post', array( (int) $r->ID ) );
                $return_value = false;
```

…is that the intent; do the code need to unschedule all existing events and reschedule them? Previously the code adjusted for daylight savings, but I don't know if it does this any longer.

Separately, could you add a docblock for this function, please?

ethitter commented:
> The core of the loop seems to be...

That's only half of it. The first bit, which checks for `false === $timestamp`, is what ensures that jobs exist for each scheduled post, which is what the plugin offers as a configuration option.

As for unscheduling and rescheduling, that was already part of this function, and mirrors what Core does in `_future_post_hook()`: https://developer.wordpress.org/reference/functions/_future_post_hook/. This ensures that the scheduled job is set for the right time, in cases where a scheduled-post is re-scheduled. Otherwise, WP would fire the job at the wrong time, the post wouldn't be transitioned because it wasn't ready, and Core wouldn't ever try to schedule it for publishing again.

As for adjusting for DST, that will still happen because the function `get_gmt_from_date()` would return a different timestamp before/after a DST switch, which would invoke the rescheduling logic. All that was really removed was the manual updating of the `post_date_gmt` column in `$wpdb->posts`, because that column is no longer used by the plugin (it's unindexed, which prompted the move to the indexed `post_date` column).

> Separately, could you add a docblock for this function, please?

Sure!

ethitter committed:
Explicitly set timezone, rather than relying on current PHP behaviour.

Since Core calls `date_default_timezone_set( 'UTC' )` in `wp-settings.php`, and `strtotime()` uses that value if no timezone is specified, this isn't strictly necessary, but it can't hurt not to assume.

simonwheatley reviewed (approved):
Looks good to me 👍
